hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/ClassDefinitionImpl.java

```diff
@@ -152,31 +149,23 @@ public boolean equals(Object o) {
             return false;
         }
         ClassDefinitionImpl that = (ClassDefinitionImpl) o;
-        if (factoryId != that.factoryId) {
-            return false;
-        }
-        if (classId != that.classId) {
-            return false;
-        }
-        if (version != that.version) {
-            return false;
-        }
-        return fieldDefinitionsMap.equals(that.fieldDefinitionsMap);
+        return portableId.equals(that.portableId)
+                && fieldDefinitionsMap.equals(that.fieldDefinitionsMap);
     }
 
     @Override
     public int hashCode() {
-        int result = classId;
-        result = 31 * result + version;
+        int result = portableId.getClassId();
```

Comment:
previous implementation ignored `factoryId` but that made it inconsistend with hashCode. If that is a feature (it is?), then maybe it should be reflected in `PortableId.hashCode`?

Reply:
This hash code is used by `DefaultSerializationServiceBuilder#classDefinitions`, `SerializationConfig#classDefinitions` and `PortableHookLoader#definitions`, which are all `HashSet`. Changing the hash code may cause the same element to be added twice, but none of the classes are serializable/persistent/shared. Even if they were serializable, `HashSet` would recompute their hashes during deserialization. Thus, the implementation can be changed to return `portableId.hashCode()`.

Reply:
Possibly `factoryId` is not very selective and excluding it from `hashCode` makes `hashCode` slightly faster for the price of potentially more collisions in the hashmap. But according to your analysis (and I agree) it should not matter for anything else.

@hazelcast/apis WDYT?

Reply:
It seems that the reason for excluding `factoryId` from keys is that each factory has its own `ClassDefinitionContext`. However, `hashCode()` shouldn't be tied to any specific usage. Furthermore, `ClassDefinition` is never used as a key, except for the usages listed above. In `PortableContextImpl.ClassDefinitionContext#versionedDefinitions`, it is used as a value, and the key is a bitwise concatenation, where high-order 32 bits are `classId` and low-order 32 bits are `version`.